### ISSUE TITLE
Bleeding characters + color-codes on split scrolling

### ISSUE THREAD

**Issue by alexozer** [neovim/neovim#3875]
It appears that scrolling a vertical split can sometimes cause characters and color-codes to "bleed out" on the left side of the terminal. [Here](https://youtu.be/SeV1MyKY9yA)'s a little demo, and [this](https://gist.github.com/9800a49c2d77d821daee) is the file I was testing with. The effect seems greater when there are either more characters in a split (with a more "dense" file and/or smaller font), when the split is scrolled more quickly, or both.

![screenshot](https://cloud.githubusercontent.com/assets/2074322/11924601/16651dca-a781-11e5-94a1-2faee7dd2307.png)

I have tested with gnome-terminal, st, and xfce4-terminal, and it occurs in all of them, so I don't think the terminal is at fault. I've also managed to reproduce this with an empty init.vim.

**Comment by synaptiko** [comment]
I've noticed similar thing recently but when switching between tabs...

![screenshot from 2016-01-08 16-31-48](https://cloud.githubusercontent.com/assets/1593280/12295847/2d7b53cc-ba04-11e5-9774-1046ab72662a.png)

**Comment by justinmk** [comment]
Please say your $TERM

**Comment by synaptiko** [comment]
$TERM=gnome-256color and I'm using gnome-terminal 3.18.2-1 on ArchLinux

**Comment by justinmk** [comment]
Does the behavior change if you `export TERM=xterm-256color` before running `nvim` ?

**Comment by alexozer** [comment]
The behavior remains the same for me when running `export TERM=xterm-256color` before `nvim` (on gnome-terminal 3.18.2-1 + Arch)

**Comment by justinmk** [comment]
@synaptiko @alexozer Do you have `$COLORTERM` set and if so can you try without it?

**Comment by alexozer** [comment]
I don't have `$COLORTERM` set.

**Comment by synaptiko** [comment]
@justinmk I also don't have $COLORTERM set and with xterm-256color it's the same.

**Comment by synaptiko** [comment]
Btw. it's not happening under Guake, but in Gnome Terminal it is.

**Comment by alexozer** [comment]
@synaptiko I was actually able to reproduce it with Guake. Have you tried it with the test file and a small font?

**Comment by synaptiko** [comment]
@alexozer Yes, with various font-sizes. I tried both `TERM=gnome-256color nvim` and `TERM=xterm-256color nvim`. So for me Guake seems to be working without any issue.

**Comment by alexozer** [comment]
@synaptiko [Here](https://www.youtube.com/watch?v=Cf9A-w-jg4c&feature=youtu.be)'s a screencast of my experience

**Comment by synaptiko** [comment]
@alexozer Ok, I have to take it back. When I open three vertical splits as it's in your screencast then it starts happening again.

**Comment by FuzzOli87** [comment]
I see the same thing that @synaptiko sees on terminator-gtk3-bz on Arch. I'm using 
```
let $NVIM_TUI_ENABLE_TRUE_COLOR=1
let $NVIM_TUI_ENABLE_CURSOR_SHAPE=1
```

I'm using oceanic-next true color scheme. No particular action triggers it. Just happens every once in a while.

**Comment by justinmk** [comment]
All terminals reported above are libvte terminals[1], including guake. What version of libvte are you using?

1. https://gist.github.com/XVilka/8346728#now-supporting-truecolour

**Comment by alexozer** [comment]
Also, in case it wasn't apparent, this behavior does not occur in normal
Vim.

**Comment by synaptiko** [comment]
@justinmk I'm using libvterm 0.1.git1.c4317a6-2.

**Comment by alexozer** [comment]
@justinmk I just tested on xterm, a non-VTE terminal, and still experienced the bug:
![image](https://cloud.githubusercontent.com/assets/2074322/12427957/01902b4c-beb0-11e5-8e66-32064a31ac0d.png)

**Comment by justinmk** [comment]
Hmm, I wonder if it's a terminfo issue. 

Is anyone _not_ having this issue on linux?

**Comment by wsdjeg** [comment]
can not reproduce,what is the step?

**Comment by crondog** [comment]
@justinmk I am using Arch linux and this does not affect me (nvim v0.1.2-265-gdddbf9c). I tested with termite (vte) and st

However in termite where ex commands go, I notice a flash of ~@k or <80> but then they disappear. The same for st but instead it is ^Y and ^E which are the codes for mouse scroll up/down. This only happens when I scroll really fast though and only sometimes

**Comment by fbernier** [comment]
I have an issue which looks similar to @synaptiko. I tried recording an asciicast, but the problem does not appear in it -_-

https://asciinema.org/a/96b3b4vlrcmc79kmmwedxcfsl

Here is a screenshot of what actually appears in the status line:

http://i.imgur.com/qMR5jV9.png

**Comment by FuzzOli87** [comment]
Started using Terminix recently and it's showing the same issues.

**Comment by FuzzOli87** [comment]
This is also happening on my mac on the nightly version of iTerm2. I'm still not sure what triggers it but it does become pretty annoying and distracting, especially when this happens on actual code lines and I'm forced to close the editor and start over. Happens too many times for comfort. 


![screen shot 2016-07-13 at 1 11 05 pm](https://cloud.githubusercontent.com/assets/12246915/16814417/62c214f4-48fb-11e6-8170-8cbc533b56ea.png)

**Comment by b1101** [comment]
- Arch linux
- Termite
- Neovim 0.1.4

It is kind of weird, but the issue appeared the same day I switched to a 4k screen. Have never encountered it before. Can it be anyhow related?

![img-2016-30-07-232020](https://cloud.githubusercontent.com/assets/12877905/17273334/6d3b4248-56b1-11e6-952b-7680264159e7.png)

**Comment by wasowski** [comment]
Notably, I am *not* seeing this problem for quite some time with Ubuntu --- presently NVIM 0.1.5-dev, gnome-terminal 3.18.3 and [this script](https://gist.github.com/wasowski/b0dc638cbd9bab4b5e4ecad074d156b6).  Don't really have time on my hands to debug why, but perhaps someone else can see what is fixing this.

The nvim-wrapper script is available online, but I include the gist to ensure that the actually working version is available here. I have not written that script.

**Comment by justinmk** [comment]
@wasowski Without the wrapper script, the problem still occurs for you?

**Comment by wasowski** [comment]
Very good point @justinmk . Just tested with the same gnome-terminal and it does not.  So the script is a red herring, ignore it.

**Comment by argentum47** [comment]
happens with me in a normal gnome-terminal as well. as well as in rxvt and sakura. I wonder what is wrong.  then again if I resize the terminal, it becomes ok, (this is rxvt).. same with the terminal wrapper.
![screenshot from 2016-08-16 10 26 25](https://cloud.githubusercontent.com/assets/6324904/17688328/ee679e1a-639b-11e6-9a79-b11baa11edca.png)

**Comment by romgrk** [comment]
On terminix: 

![screenshot from 2016-08-20 02-35-30](https://cloud.githubusercontent.com/assets/1423607/17829520/2538f494-667f-11e6-9c1a-ebf6f5734ebd.png)

**Comment by raichoo** [comment]
I've been living with this for approximately a year now and it's getting a bit frustrating… is there any more information that can be provided to pin this down?

**Comment by justinmk** [comment]
@raichoo Did you try different terminals or maybe a different `$TERM`? This is most likely a bad interaction with unibilium or a bug in our screen code.

A reduced test case would help a lot. So far I see a bunch of screenshots of a problem I've never encountered on all kinds of systems. Except I don't use gnome.

**Comment by raichoo** [comment]
@justinmk There is a way I can reproduce this: I'm using `gnome-terminal`. `TERM=xterm-256color`.

1. open terminal
2. switch terminal to gnome full screen
3. open a highly colorful file (e.g. CMakeLists.txt from the `neovim` tree)
4. do a vertical split so the same file shows up in the split
5. stay in the left split
6. press `C-l`
7. if no strange character appears in the right split go to the next line. goto 6.

I've also compiled a minimal `init.vim` to reproduce the error.

```viml

set rtp+=~/.nvim/plugged/molokai
set termguicolors
set listchars=tab:▸\ ,eol:¬,extends:❯,precedes:❮,nbsp:·
set number
set list

color molokai

nnoremap <silent> <C-l> :noh<CR><C-l>
```
It seems to be a combination of these features that trigger the bug. At least on my machine.

**Comment by justinmk** [comment]
So without the listchars setting it does not happen? Is everyone else here using a Unicode listchars setting?

**Comment by raichoo** [comment]
I've tried it with a non-unicide `eol` listchar and the bug didn't occur. I'm not sure if this is what triggers the bug for everyone else, I'm not even sure if this is the only configuration that triggers it, but it's perfectly reproducible on my machine.

**Comment by justinmk** [comment]
Perhaps related to drawing Unicode cells in general. Some length calculation may be off.

**Comment by raichoo** [comment]
That sounds like a possibility. I also see unicode listchars in the screenshot that @romgrk made.

**Comment by raichoo** [comment]
Hm, I've tried working with `number`, `list` and `listchars` turned off and the bug still showed up. I guess we can rule out `listchars` as well as `number`.

**Comment by raichoo** [comment]
I have managed to reproduce the bug with a file that heavily uses Unicode while using the minimal `vimrc` with `listchars`, `list` and `number` turned off.

```viml

syntax on
set rtp+=~/.nvim/plugged/molokai
set termguicolors
"set listchars=eol:¬
" set number
"set list

color molokai

nnoremap <silent> <C-l> :noh<CR><C-l>
```
![screenshot from 2016-08-30 13-42-30](https://cloud.githubusercontent.com/assets/151506/18087943/e8e63ad8-6eb7-11e6-99f0-b1dbb4661204.png)

I've opened a bunch of splits set the filetype to `haskell` and moved around pressing `C-l` on every one. This configuration resulted on the weird character in the right split when I pressed `C-l` on line 7 in the left split.
I don't know if this is related to Unicode though, it might just lead to the bug being triggered more often.

**Comment by raichoo** [comment]
I could also reproduce this error using the `desert` colorscheme, haven't managed to do so with `default`.

**Comment by raichoo** [comment]
I've checked this with an ASCII file and I could reproduce it as well, so this seems to be pointing to a problem with `termguicolors` and more elaborate colorschemes.

**Comment by romgrk** [comment]
To extend the minimal test case described above, if the following line in the `desert.vim` colorscheme
```viml
hi Normal	guifg=White guibg=grey20
```
is changed to 
```viml
hi Normal	guifg=White
```
the issue disappears.

**Comment by raichoo** [comment]
I've added `highlight Normal guifg=White guibg=None` to my `init.vim` and I haven't the issue for a day now (which is a really good sign). Maybe it just decreases the chance of the bug showing up but I will continue to monitor this over the course of the next week.

**Comment by laerus** [comment]
getting this as well, it seems pretty random although this happens mostly when i have nvim opened in terminator in the far left side of the screen.
![neovim-bug](https://cloud.githubusercontent.com/assets/6930157/18555013/e2711bdc-7b6e-11e6-9711-6f7ad8c95782.png)

-

![screenshot from 2016-09-15 14-40-48](https://cloud.githubusercontent.com/assets/6930157/18555019/e4c62d64-7b6e-11e6-9a1e-7f50e7c276b1.png)

opening ex mode and hitting backspace makes this worse.

tried removing all plugins, (with same init.vim) and this still happens.

**Comment by sherzinger** [comment]
I'm getting this as well, but only on external displays. (s. closed issue above)
The workaround does not work for me.

**Comment by laerus** [comment]
it seems i'm not getting this anymore after removing this from init.vim
`highlight ColorColumn ctermbg=80`
but again this maybe just luck cause this issue happens pretty randomly for me.

**Comment by YorickPeterse** [comment]
I ran into this as well, but only a few times. In my case Neomake adds a bunch of Unicode characters on lines that contain warnings and such. Switching those characters over to pure ASCII characters thus far seems to solve the problem.

**Comment by justinmk** [comment]
#5436 may help this issue, if someone here can try latest master.

**Comment by sherzinger** [comment]
I can confirm that the issue is solved for me.
Awesome, thanks!

**Comment by justinmk** [comment]
Closing this. Will re-open if someone reports differently.

**Comment by allyraza** [comment]
I am also having this problem

**Comment by curvi** [comment]
I have these bugs for way more than a year. I use the most recent Ubuntu build 0.2.1-dev and the most recent fedora build both on Gnome Terminal and get these fragments and "bleeding" on various Laptops/Display/Linux combinations!
It seems to have something to do with fullscreen/window resize, other than that I cannot specify as much as I would like.
Please fix this - this makes nvim a real bummer to use!

**Comment by justinmk** [comment]
@curvi This issue is not a catch-all for general complaints about display issues.

**Comment by sherzinger** [comment]
@justinmk Is there open issue related to this one or can we maybe re-open it? 
This bleeding appears again quite frequently for me. It's especially notable in large files when scrolling via <C-d> and <C-u>

(Using v0.2.2 on macOS 10.13.1)

**Comment by alefa** [comment]
I'm experiencing this issue as well (i.e. bleeding characters when scrolling in a vertical split). It's pretty annoying. If there's no other open bug report on this issue, can we reopen this one as @sherzinger suggested?
I'm on Debian 9 with Neovim 0.2.2 compiled from source.

**Comment by justinmk** [comment]
@alefa This issue regressed in 0.2.3, it's fixed again on 0.2.3 (the development version). See https://github.com/neovim/neovim/pull/7676

The [Releases](https://github.com/neovim/neovim/releases) page has pre-built archives for Linux/Windows/macOS. (No installation needed, just unzip and run. For Linux, [nvim.appimage](https://github.com/neovim/neovim/releases/download/nightly/nvim.appimage) is a universal "binary", just `chmod u+x nvim.appimage && ./nvim.appimage`)
